pages/dictionary/index.tsx

```diff
@@ -24,25 +27,104 @@ const normalizeForSearch = (text: string) => {
+  const [flattenTranslations, setFlattenTranslations] = useState<TranslationType[]>();
+  const [filteredTranslations, setFilteredTranslations] = useState<TranslationType[] | []>([]);
+  const [maxItems, setMaxItems] = useState(20);
+  const [isSticky, setIsSticky] = useState(false);
+
+  const { t, i18n } = useTranslation();
   const { currentLanguage } = useLanguage();
 
-  // filter category name and translations by search input
-  const filterBySearch = ({ category_name_cz, category_name_ua, translations }: TranslationsType) => {
-    const UACategoryName = normalizeForSearch(category_name_ua);
-    const CZCategoryName = normalizeForSearch(category_name_cz);
-
-    const searchText = normalizeForSearch(search);
-    // filter category name
-    const matchesCategoryTitle = CZCategoryName.includes(searchText) || UACategoryName.includes(searchText);
-    // filter translations
-    const matchesTranslations = translations.filter(({ cz_translation, ua_translation }) => {
-      return normalizeForSearch(cz_translation).includes(searchText) || normalizeForSearch(ua_translation).includes(searchText);
+  const lastContainerRef = useRef<HTMLDivElement | null>(null);
+  const searchContainer = useRef<HTMLDivElement | null>(null);
+  const searchButton = useRef<HTMLButtonElement | null>(null);
+  const translationsContainerRef = useRef<HTMLDivElement | null>(null);
+
+  // scrolls to top whenever user type in search input
+  useEffect(() => {
+    if (!search.trim() || window.scrollY === 0) return;
+    window.scrollTo({
+      top: 0,
     });
+  }, [search]);
+
+  //loads  more items once viewport is close enough to last element
+  useEffect(() => {
+    const lastTranslation = lastContainerRef.current;
+
+    const options = {
+      rootMargin: '100px',
+      threshold: [0],
+    };
+    const observer = new IntersectionObserver(([entry]) => {
+      const { intersectionRatio } = entry;
+      if (intersectionRatio > 0 && intersectionRatio < 1) {
+        setMaxItems((prev) => prev + 20);
+      }
+    }, options);
+
+    lastTranslation && observer.observe(lastTranslation);
+
+    return () => {
+      lastTranslation && observer.unobserve(lastTranslation);
+    };
+  });
+
+  // track when search input becomes sticky to apply styles
+  useEffect(() => {
+    const element = searchContainer.current;
+    const observer = new IntersectionObserver(
+      ([e]) => {
+        if (e.intersectionRatio < 1) {
+          setIsSticky(true);
+        } else if (e.intersectionRatio >= 1) {
+          setIsSticky(false);
+        }
+      },
+      { threshold: [1], rootMargin: '-57px 0px 0px 0px' }
+    );
+
+    element && observer.observe(element);
 
-    return matchesCategoryTitle || matchesTranslations.length > 0;
+    return () => {
+      element && observer.unobserve(element);
+    };
+  });
+
+  // filters translation based on user search
+  const filterTranslations = () => {
+    const res = flattenTranslations?.reduce(
```

Comment:
Why `reduce`? Wouldn't `filter` be simpler?

Reply:
I need to have access to already filtered results to check duplicates. For example "Ahoj" will display two results.

Reply:
But maybe I should do it on initial render